scylla-cdc/src/checkpoints.rs

```diff
@@ -0,0 +1,352 @@
+            .query(format!("SELECT * FROM {}", table), ())
+            .await
+            .unwrap()
+            .rows()
+            .unwrap()
+            .into_typed::<(StreamID, chrono::Duration, chrono::Duration)>()
+            .map(|x| x.unwrap())
+            .map(|(id, gen, time)| -> Checkpoint {
+                Checkpoint {
+                    stream_id: id,
+                    generation: GenerationTimestamp { timestamp: gen },
+                    timestamp: time.to_std().unwrap(),
+                }
+            })
+            .collect::<Vec<Checkpoint>>()
+    }
+
+    #[tokio::test]
+    async fn test_save_checkpoint_multiple_times() {
+        const N: usize = 20;
+        let (session, table_name, cp_saver) = setup().await;
+
+        let mut checkpoint = Checkpoint {
+            timestamp: Duration::from_secs(128),
+            stream_id: StreamID {
+                id: vec![1, 1, 1, 1, 1, 1, 1, 1],
+            },
+            generation: GenerationTimestamp {
+                timestamp: chrono::Duration::min_value(),
+            },
+        };
+
+        let delta = Duration::from_secs(32);
+
+        for _ in 0..N {
+            checkpoint.timestamp += delta;
+            cp_saver.save_checkpoint(&checkpoint).await.unwrap();
+        }
+
+        let checkpoints = get_checkpoints(&session, &table_name).await;
+
+        assert_eq!(checkpoints.len(), 1); // Constant stream id for every query.
+
+        let saved_checkpoint = &checkpoints[0];
+        assert_eq!(*saved_checkpoint, checkpoint);
+
+        assert_eq!(
+            checkpoint.timestamp,
+            cp_saver
+                .load_last_checkpoint(&checkpoint.stream_id)
+                .await
+                .unwrap()
+                .unwrap()
+                .to_std()
+                .unwrap()
+        )
+    }
+
+    #[tokio::test]
+    async fn test_save_generation_multiple_times() {
+        const N: usize = 20;
+        let (session, table_name, cp_saver) = setup().await;
+
+        let mut generation = GenerationTimestamp {
+            timestamp: chrono::Duration::seconds(0),
+        };
+
+        let delta = chrono::Duration::seconds(10);
+
+        for _ in 0..N {
+            generation.timestamp = generation.timestamp.add(delta);
+            cp_saver.save_new_generation(&generation).await.unwrap();
+        }
+
+        let checkpoints = get_checkpoints(&session, &table_name).await;
+
+        assert_eq!(checkpoints.len(), 1); // Constant stream id for every query.
+
+        let saved_checkpoint = &checkpoints[0];
+        assert_eq!(saved_checkpoint.generation, generation);
+
+        assert_eq!(
+            cp_saver.load_last_generation().await.unwrap().unwrap(),
+            generation
+        );
+    }
+
+    #[tokio::test]
+    async fn test_save_checkpoint_multiple_streams() {
+        const N_OF_IDS: u8 = 200;
+
+        let (_, _, cp_saver) = setup().await;
+
+        let mut checkpoints = Vec::<Checkpoint>::new();
```

Comment:
You can allocate required memory using `with_capacity` and later store elements in corresponding fields instead of pushing.